README.md
# text-metrics

[![NPM version][npm-image]][npm-url] [![Build Status][travis-image]][travis-url] [![BrowserStack Status][browserstack-image]][browserstack-url] [![Dependency Status][depstat-image]][depstat-url] [![Download][dlcounter-image]][dlcounter-url] [![Coverage Status][coveralls-image]][coveralls-url]

> An lightweight & efficient text measurement set for the browser using canvas to prevent layout reflows.

## Features

- Compute width
- Compute height
- Compute linebreaks
- Compute max font-size to fit into element

## Installation

If you're using node, you can run `npm install text-metrics`.

text-metrics is also available via [Bower](https://github.com/bower/bower) (`bower install text-metrics`)

Alternatively if you just want to grab the file yourself, you can download either the current stable [production version][min] or the [development version][max] directly.

[min]: https://raw.github.com/bezoerb/text-metrics/master/dist/text-metrics.min.js
[max]: https://raw.github.com/bezoerb/text-metrics/master/dist/text-metrics.js

## Setting it up

text-metrics supports AMD (e.g. RequireJS), CommonJS (e.g. Node.js) and direct usage (e.g. loading globally with a &lt;script&gt; tag) loading methods.
You should be able to do nearly anything, and then skip to the next section anyway and have it work. Just in case though, here's some specific examples that definitely do the right thing:

### CommonsJS (e.g. Node)

text-metrics needs some browser environment to run.

```javascript
const textMetrics = require('text-metrics');

const el = document.querySelector('h1');
const metrics = textMetrics.init(el);

metrics.width('unicorns');
// -> 210

metrics.height('Some long text with automatic word wraparound');
// -> 180
metrics.lines('Some long text with automatic word wraparound');
// -> ['Some long text', 'with automatic', 'word', 'wraparound']
metrics.maxFontSize('Fitting Headline');
// -> 33px
```

### Webpack / Browserify

```javascript
const textMetrics = require('text-metrics');
textMetrics.init(document.querySelector('.textblock')).lines();
```

### AMD (e.g. RequireJS)

```javascript
define(['text-metrics'], function(textMetrics) {
  textMetrics.init(document.querySelector('h1')).width('unicorns');
});
```

### Directly in your web page:

```html
<script src="text-metrics.min.js"></script>
<script>
  textMetrics.init(document.querySelector('h1')).width('unicorns');
</script>
```

## API

Construct textmetrics object:

`textMetrics.init([el, overwrites])`

You can call textMetrics with either an HTMLElement or with an object with style overwrites or with both.
e.g.

```javascript
// takes styles from h1
textMetrics.init(document.querySelector('h1'));

// takes styles from h1 and overwrites font-size
textMetrics.init(document.querySelector('h1'), { fontSize: '20px'});

// only use given styles
textMetrics.init({
    fontSize': '14px',
    lineHeight': '20px',
    fontFamily': 'Helvetica, Arial, sans-serif',
    fontWeight': '400',
    width: 100
});
```

## Methods

`width([text, [options, [overwrites]]])`<br/>
`height([text, [options, [overwrites]]])`<br/>
`lines([text, [options, [overwrites]]])`<br/>
`maxFontSize([text, [options, [overwrites]]])`<br/>

#### text

Type: `string`
Defaults to `el.innerText` if an element is available

#### options

Type: `object`

| key       | default | description                                                        |
| --------- | ------- | ------------------------------------------------------------------ |
| multiline | `false` | The width of widest line instead of the width of the complete text |

#### overwrites

Type: `object`

Use to overwrite styles

## Performance

I've compared this module with a very naive jQuery implementation as well as
the . See https://jsperf.com/bezoerb-text-metrics
Even if `Range.getBoundingClientRect` should be considered as a performance bottleneck according to
[what-forces-layout](https://gist.github.com/paulirish/5d52fb081b3570c81e3a) by Paul Irish,
i couldn't detect any sort of recalculate style and it massively outperforms `textMetrics.height()`.

## Compatibility

The normal build (3,2kb gzipped) should work well on modern browsers. There's also a compatibility build (10,7kb) available if you need to support IE11.<br/>
These builds are tested using <br/><a href="http://browserstack.com/" target="_blank"><img src="./test/fixtures/Browserstack-logo.svg" width="200" alt="Browserstack"></a>

## License

Copyright (c) 2016 Ben Zörb
Licensed under the [MIT license](http://bezoerb.mit-license.org/).

[npm-url]: https://npmjs.org/package/text-metrics
[npm-image]: https://badge.fury.io/js/text-metrics.svg
[travis-url]: https://travis-ci.org/bezoerb/text-metrics
[travis-image]: https://secure.travis-ci.org/bezoerb/text-metrics.svg?branch=master
[depstat-url]: https://david-dm.org/bezoerb/text-metrics
[depstat-image]: https://david-dm.org/bezoerb/text-metrics.svg
[dlcounter-url]: https://www.npmjs.com/package/text-metrics
[dlcounter-image]: https://img.shields.io/npm/dm/text-metrics.svg
[coveralls-url]: https://coveralls.io/github/bezoerb/text-metrics?branch=master
[coveralls-image]: https://coveralls.io/repos/github/bezoerb/text-metrics/badge.svg?branch=master
[browserstack-url]: https://www.browserstack.com/automate/public-build/UFVIM0xVaHpucm9YVUN4TXJtYStvQUFWSEFWMFFiQVhaNGZPaXJ6dWNsTT0tLXdnNTU0OGtDNWhIRVJ4SFBhZkFMeWc9PQ==--d7c5f52dc2b2c9bcb2cf1dce052e84741e37f037
[browserstack-image]: https://www.browserstack.com/automate/badge.svg?badge_key=UFVIM0xVaHpucm9YVUN4TXJtYStvQUFWSEFWMFFiQVhaNGZPaXJ6dWNsTT0tLXdnNTU0OGtDNWhIRVJ4SFBhZkFMeWc9PQ==--d7c5f52dc2b2c9bcb2cf1dce052e84741e37f037

## License

[MIT](https://bezoerb.mit-license.org/) © [Ben Zörb](http://sommerlaune.com)


CHANGELOG.md
# v1.0.1 / 2019-06-13

==================

- Bump dependencies

# v1.0.0 / 2019-01-22

- rewrite
- drop node 6 support
- add node 10 test
- minor tweaks
- Bump dependencies
- Bump dependencies & tweaked tests

# v0.4.3 / 2017-05-02

- fixed line calculation

# v0.4.2 / 2017-05-02

- Bump dependencies
- Fixed some rounding issues

# v0.4.1 / 2017-02-13

- Minor tweaks & readme
- Add .eslintrc for codacy

# v0.4.0 / 2017-02-09

- Dropped full htmlentity decode in favor of filesize

# v0.3.1 / 2017-02-09

- Performance improvements for 'height()'

# v0.3.0 / 2017-02-07

- Use innerText to see <br/>

# v0.2.0 / 2017-02-06

- 0.2.0
- Add array.includes polyfill
- switch to babel-preset-env
- Consider linebreak code points

# v0.1.4 / 2016-12-12

- 0.1.4
- Trim text

# v0.1.3 / 2016-12-12

- 0.1.3
- Trim text

# v0.1.2 / 2016-12-12

- 0.1.2
- minor fix

# v0.1.1 / 2016-12-12

- 0.1.1
- readme
- readme
- readme
- pass style overwrites to lines()
- keep node 5 compatibility

# v0.1.0 / 2016-12-11

- 0.1.0
- consider letter-spacing & word-spacing

# v0.0.2 / 2016-12-09

- 0.0.2
- linter
- build
- build
- build
- build
- 0.0.1

# v0.0.1 / 2016-12-08

- 0.0.1
- initial commit


test/test.js
/* eslint-env es6, browser */
// eslint-disable-next-line import/no-unassigned-import
import '../node_modules/babel-core/register';
import test from 'ava';
import {getContext2d} from '../src/utils';
import textMetrics from '../src';

test('Compute with without element', t => {
    const expected = getContext2d(
        '400 30px Helvetica, Arial, sans-serif'
    ).measureText('test');

    const val = textMetrics({
        'font-size': '30px',
        'font-weight': '400',
        'font-family': 'Helvetica, Arial, sans-serif'
    }).width('test');
    t.is(val, expected.width);
});

test('Compute width for h1', t => {
    const expected = getContext2d(
        '500 36px Helvetica, Arial, sans-serif'
    ).measureText('TEST');

    const el = document.querySelector('h1');
    const val = textMetrics(el).width('TEST');
    t.is(val, expected.width);
});

test('Compute width for h2', t => {
    // Lowercase as this get's applied via css
    const expected = getContext2d(
        '500 30px Helvetica, Arial, sans-serif'
    ).measureText('test');

    const el = document.querySelector('h2');
    const val = textMetrics(el).width('TEST');
    t.is(val, expected.width);
});

test('Computes width', t => {
    const el = document.querySelector('h1');
    const metrics = textMetrics(el);
    const v1 = metrics.width('-');
    const v2 = metrics.width('--');
    const v3 = metrics.width('---');

    t.true(v1 < v2 < v3);
});

test('Computes maxFontSize', t => {
    const el = document.querySelector('#max-font-size');
    const val = textMetrics(el).maxFontSize('unicorn');
    t.is(val, '183px');
});

test('Computes lines', t => {
    const el = document.querySelector('#height');

    const text =
        'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aliquam atque cum dolor explicabo incidunt.';
    const expected = [
        'Lorem ipsum',
        'dolor sit amet,',
        'consectetur',
        'adipisicing elit.',
        'Aliquam atque',
        'cum dolor',
        'explicabo',
        'incidunt.'
    ];

    const value = textMetrics(el).lines(text);

    t.is(value.length, expected.length);

    for (let i = 0; i < value.length; i++) {
        t.is(value[i], expected[i]);
    }
});

test('Computes lines with breaks', t => {
    const el = document.querySelector('#lines');
    const text =
        'Lo&shy;rem ipsum d&shy;o&shy;lor sit amet, c&mdash;onsectur a&mdash;dipisicing elit. Aliquam atque cum dolor explicabo &bigstar;.';
    const expected = [
        'Lorem ipsum do-',
        'lor sit amet, c—',
        'onsectur a—',
        'dipisicing elit.',
        'Aliquam atque',
        'cum dolor',
        'explicabo',
        '&bigstar;.'
    ];

    const value = textMetrics(el).lines(text);

    t.is(value.length, expected.length);

    for (let i = 0; i < value.length; i++) {
        t.is(value[i], expected[i]);
    }
});

test('Computes lines with break-all', t => {
    const el = document.querySelector('#lines-break');
    // Need to pass text as the jsdom implementation of innerText differs from browsers
    // https://github.com/tmpvar/jsdom/issues/1245
    const text =
        'Lorem ipsum d&shy;o&shy;lor sit amet, c&mdash;onsectur a&mdash;dipisicing elit. Ali<br/>quam atque cum dolor explicabo incidunt.';
    const expected = [
        'Lorem ipsum d-',
        'olor sit amet, c',
        '—onsectur a—',
        'dipisicing elit. Al',
        'i',
        'quam atque cu',
        'm dolor explica',
        'bo incidunt.'
    ];

    const value = textMetrics(el).lines(text);
    t.is(value.length, expected.length);

    for (let i = 0; i < value.length; i++) {
        t.is(value[i], expected[i]);
    }
});

test('Computes height', t => {
    const text =
        'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aliquam atque cum dolor explicabo incidunt.';

    const val = textMetrics({
        'font-size': '14px',
        'line-height': '20px',
        'font-family': 'Helvetica, Arial, sans-serif',
        'font-weight': '400',
        width: 100
    }).height(text);

    t.is(val, 160);
});

test('Consider letter- and word-spacing', t => {
    const referenceline = 'sit amet, consectetur';
    const reference = getContext2d(
        '500 24px Helvetica, Arial, sans-serif'
    ).measureText(referenceline);
    // Width + word-spacing (10px) + letter-spacing (2px)
    // eslint-disable-next-line  no-mixed-operators
    const referenceWidth = reference.width + 20 + referenceline.length * 2;

    const el = document.querySelector('h3');
    const metrics = textMetrics(el, {'line-height': '26px'});

    const lines = metrics.lines();
    const width = metrics.width(null, {multiline: true});
    const height = metrics.height();

    const expected = [
        'Lorem ipsum dolor',
        'sit amet, consectetur',
        'adipisicing elit.',
        'Aliquam atque cum',
        'dolor explicabo',
        'incidunt.'
    ];

    t.is(lines.length, expected.length);

    for (let i = 0; i < lines.length; i++) {
        t.is(lines[i], expected[i]);
    }

    t.is(width, referenceWidth);
    t.is(height, 26 * expected.length);
});

test('Consider multiple line breaking characters', t => {
    const text =
        'Von Kabeljau über Lachs und Thunfisch bis hin zu Zander – unsere Fisch-Vielfalt wird Sie begeistern. Bestimmt!';
    const lines = textMetrics({
        'font-size': '14px',
        'line-height': '20px',
        'font-family': 'Helvetica, Arial, sans-serif',
        'font-weight': '400',
        width: 400
    }).lines(text);

    const expected = [
        'Von Kabeljau über Lachs und Thunfisch bis hin zu Zander –',
        'unsere Fisch-Vielfalt wird Sie begeistern. Bestimmt!'
    ];

    t.is(lines.length, expected.length);
    for (let i = 0; i < lines.length; i++) {
        t.is(lines[i], expected[i]);
    }
});

test('Computes lines with one very long word', t => {
    const el = document.querySelector('#height');
    const text =
        'Craspharetrapharetragravida.Vivamusconsequatlacusvelposuerecongue.Duisaloremvitaeexauctorscelerisquenoneuturpis.Utimperdietmagnasitametjustobibendumvehicula.';
    const expected = [
        'Craspharetraph',
        'aretragravida.Vi',
        'vamusconsequ',
        'atlacusvelposu',
        'erecongue.Duis',
        'aloremvitaeexa',
        'uctorscelerisqu',
        'enoneuturpis.Ut',
        'imperdietmagn',
        'asitametjustobi',
        'bendumvehicul',
        'a.'
    ];

    const value = textMetrics(el).lines(text, {}, {'word-break': 'break-all'});

    t.is(value.length, expected.length);

    for (let i = 0; i < value.length; i++) {
        t.is(value[i], expected[i]);
    }
});


test/utils.js
/* eslint-env es6, browser */
// eslint-disable-next-line import/no-unassigned-import
import '../node_modules/babel-core/register';
import test from 'ava';
import {CSSStyleDeclaration} from 'cssstyle';
import * as utils from '../src/utils';

test('isElement', t => {
    const el = document.querySelector('h1');

    t.false(utils.isElement({}));
    t.false(utils.isElement(1));
    t.false(utils.isElement(false));
    t.false(utils.isElement(null));
    t.false(utils.isElement(undefined));
    t.true(utils.isElement(el));
});

test('isObject', t => {
    t.true(utils.isObject({}));
    t.false(utils.isObject(1));
    t.false(utils.isObject(false));
    t.false(utils.isObject(null));
    t.false(utils.isObject(undefined));
    t.false(utils.isObject([]));
});

test('isCSSStyleDeclaration', t => {
    const style = new CSSStyleDeclaration();

    t.true(utils.isCSSStyleDeclaration(style));
    t.false(utils.isCSSStyleDeclaration({}));
});

test('getStyle', t => {
    const el = document.querySelector('h1');
    const style = utils.getStyle(el);

    t.true(typeof style.getPropertyValue === 'function');
    t.true(utils.isCSSStyleDeclaration(style));
});

test('getFont', t => {
    const style = new CSSStyleDeclaration();
    style.setProperty('font-family', 'Helvetica, Arial');
    style.setProperty('font-size', '1em');
    style.setProperty('font-style', 'italic');
    style.setProperty('font-weight', 'bolder');

    t.is(utils.getFont(style), 'bolder italic 16px Helvetica, Arial');

    const el = document.querySelector('h1');
    t.is(utils.getFont(utils.getStyle(el)), '500 36px Helvetica, Arial, sans-serif');
});

test('canGetComputedStyle', t => {
    const el = document.querySelector('h1');

    t.true(utils.canGetComputedStyle(el));
    t.false(utils.canGetComputedStyle({}));
});

test('getStyledText', t => {
    const el = document.querySelector('h1');
    const style = utils.getStyle(el);
    const text = 'This is a test string';

    t.is(utils.getStyledText(text, style), 'THIS IS A TEST STRING');
});

test('prop', t => {
    const obj = {a: 42, b: '42'};

    t.is(utils.prop(obj, 'a', 0), 42);
    t.is(utils.prop(obj, 'b'), '42');
    t.is(utils.prop(obj, 'c', 1), 1);
    t.not(utils.prop(obj, 'c'), 1);
});

test('normalizeOptions', t => {
    const obj = {a: 1, 'font-size': '1px', lineHeight: 1};
    const normalized = utils.normalizeOptions(obj);

    t.is(normalized.a, 1);
    t.is(normalized['font-size'], '1px');
    t.is(normalized['line-height'], 1);
    t.falsy(normalized.lineHeight);
});


rollup.config.js
import nodeResolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import babel from 'rollup-plugin-babel';
import replace from 'rollup-plugin-replace';
import {terser} from 'rollup-plugin-terser';

import pkg from './package.json';

export default [
  // CommonJS
  {
    input: 'src/index.js',
    output: {file: 'lib/text-metrics.js', format: 'cjs', indent: false},
    external: [...Object.keys(pkg.dependencies || {}), ...Object.keys(pkg.peerDependencies || {})],
    plugins: [babel()],
  },

  // ES
  {
    input: 'src/index.js',
    output: {file: 'es/text-metrics.js', format: 'es', indent: false},
    external: [...Object.keys(pkg.dependencies || {}), ...Object.keys(pkg.peerDependencies || {})],
    plugins: [babel()],
  },

  // ES for Browsers
  {
    input: 'src/index.js',
    output: {file: 'es/text-metrics.mjs', format: 'es', indent: false},
    plugins: [
      nodeResolve({
        jsnext: true,
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('production'),
      }),
      terser({
        compress: {
          pure_getters: true,
          unsafe: true,
          unsafe_comps: true,
          warnings: false,
        },
      }),
    ],
  },

  // UMD Development
  {
    input: 'src/index.js',
    output: {
      file: 'dist/text-metrics.js',
      format: 'umd',
      name: 'textMetrics',
      indent: false,
    },
    plugins: [
      nodeResolve({
        jsnext: true,
      }),
      babel({
        exclude: 'node_modules/**',
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('development'),
      }),
    ],
  },

  // UMD Production
  {
    input: 'src/index.js',
    output: {
      file: 'dist/text-metrics.min.js',
      format: 'umd',
      name: 'textMetrics',
      indent: false,
    },
    plugins: [
      nodeResolve({
        jsnext: true,
      }),
      babel({
        exclude: 'node_modules/**',
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('production'),
      }),
      terser({
        compress: {
          pure_getters: true,
          unsafe: true,
          unsafe_comps: true,
          warnings: false,
        },
      }),
    ],
  },

  // UMD Development (compatibility)
  {
    input: 'src/index.compat.js',
    output: {
      file: 'dist/text-metrics.compat.js',
      format: 'umd',
      name: 'textMetrics',
      indent: false,
    },
    plugins: [
      nodeResolve({
        jsnext: true,
        main: true,
        browser: true,
      }),
      commonjs({include: 'node_modules/**'}),
      babel({
        exclude: 'node_modules/**',
        babelrc: false,
        presets: [
          [
            '@babel/env',
            {
              targets: {
                ie: '11',
              },
              modules: false,
              useBuiltIns: 'usage', //enables babel and import babel into your inputFile.js
            },
          ],
        ],
        plugins: ['@babel/proposal-object-rest-spread'],
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('development'),
      }),
    ],
  },

  // UMD Production (compatibility)
  {
    input: 'src/index.js',
    output: {
      file: 'dist/text-metrics.compat.min.js',
      format: 'umd',
      name: 'textMetrics',
      indent: false,
    },
    plugins: [
      nodeResolve({
        jsnext: true,
        main: true,
        browser: true,
      }),
      commonjs({include: 'node_modules/**'}),
      babel({
        exclude: 'node_modules/**',
        babelrc: false,
        presets: [
          [
            '@babel/env',
            {
              targets: {
                ie: '11',
              },
              modules: false,
              useBuiltIns: 'usage', //enables babel and import babel into your inputFile.js
            },
          ],
        ],
        plugins: ['@babel/proposal-object-rest-spread'],
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('production'),
      }),
      terser({
        compress: {
          pure_getters: true,
          unsafe: true,
          unsafe_comps: true,
          warnings: false,
        },
      }),
    ],
  },
];


src/utils.js
/*
 B2	Break Opportunity Before and After	Em dash	Provide a line break opportunity before and after the character
 BA	Break After	Spaces, hyphens	Generally provide a line break opportunity after the character
 BB	Break Before	Punctuation used in dictionaries	Generally provide a line break opportunity before the character
 HY	Hyphen	HYPHEN-MINUS	Provide a line break opportunity after the character, except in numeric context
 CB	Contingent Break Opportunity	Inline objects	Provide a line break opportunity contingent on additional information
 */

// B2 Break Opportunity Before and After - http://www.unicode.org/reports/tr14/#B2
const B2 = ['\u2014'];

const SHY = [
  // Soft hyphen
  '\u00AD',
];

// BA: Break After (remove on break) - http://www.unicode.org/reports/tr14/#BA
const BAI = [
  // Spaces
  '\u0020',
  '\u1680',
  '\u2000',
  '\u2001',
  '\u2002',
  '\u2003',
  '\u2004',
  '\u2005',
  '\u2006',
  '\u2008',
  '\u2009',
  '\u200A',
  '\u205F',
  '\u3000',
  // Tab
  '\u0009',
  // ZW Zero Width Space - http://www.unicode.org/reports/tr14/#ZW
  '\u200B',
  // Mandatory breaks not interpreted by html
  '\u2028',
  '\u2029',
];

const BA = [
  // Hyphen
  '\u058A',
  '\u2010',
  '\u2012',
  '\u2013',
  // Visible Word Dividers
  '\u05BE',
  '\u0F0B',
  '\u1361',
  '\u17D8',
  '\u17DA',
  '\u2027',
  '\u007C',
  // Historic Word Separators
  '\u16EB',
  '\u16EC',
  '\u16ED',
  '\u2056',
  '\u2058',
  '\u2059',
  '\u205A',
  '\u205B',
  '\u205D',
  '\u205E',
  '\u2E19',
  '\u2E2A',
  '\u2E2B',
  '\u2E2C',
  '\u2E2D',
  '\u2E30',
  '\u10100',
  '\u10101',
  '\u10102',
  '\u1039F',
  '\u103D0',
  '\u1091F',
  '\u12470',
];

// BB: Break Before - http://www.unicode.org/reports/tr14/#BB
const BB = ['\u00B4', '\u1FFD'];

// BK: Mandatory Break (A) (Non-tailorable) - http://www.unicode.org/reports/tr14/#BK
const BK = ['\u000A'];

/* eslint-env es6, browser */
const DEFAULTS = {
  'font-size': '16px',
  'font-weight': '400',
  'font-family': 'Helvetica, Arial, sans-serif',
};

/**
 * We only support rem/em/pt conversion
 * @param val
 * @param options
 * @return {*}
 */
function pxValue(val, options) {
  if (!options) {
    options = {};
  }

  const baseFontSize = parseInt(prop(options, 'base-font-size', 16), 10);

  const value = parseFloat(val);
  const unit = val.replace(value, '');
  // eslint-disable-next-line default-case
  switch (unit) {
    case 'rem':
    case 'em':
      return value * baseFontSize;
    case 'pt':
      return value / (96 / 72);
    case 'px':
      return value;
  }

  throw new Error('The unit ' + unit + ' is not supported');
}

/**
 * Get computed word- and letter spacing for text
 * @param ws
 * @param ls
 * @return {function(*)}
 */
export function addWordAndLetterSpacing(ws, ls) {
  const blacklist = ['inherit', 'initial', 'unset', 'normal'];

  let wordAddon = 0;
  if (ws && blacklist.indexOf(ws) === -1) {
    wordAddon = pxValue(ws);
  }

  let letterAddon = 0;
  if (ls && blacklist.indexOf(ls) === -1) {
    letterAddon = pxValue(ls);
  }

  return text => {
    const words =
      text
        .trim()
        .replace(/\s+/gi, ' ')
        .split(' ').length - 1;
    const chars = text.length;

    return words * wordAddon + chars * letterAddon;
  };
}

/**
 * Map css styles to canvas font property
 *
 * font: font-style font-variant font-weight font-size/line-height font-family;
 * http://www.w3schools.com/tags/canvas_font.asp
 *
 * @param {CSSStyleDeclaration} style
 * @param {object} options
 * @returns {string}
 */
export function getFont(style, options) {
  const font = [];

  const fontWeight = prop(options, 'font-weight', style.getPropertyValue('font-weight')) || DEFAULTS['font-weight'];
  if (
    ['normal', 'bold', 'bolder', 'lighter', '100', '200', '300', '400', '500', '600', '700', '800', '900'].indexOf(
      fontWeight.toString()
    ) !== -1
  ) {
    font.push(fontWeight);
  }

  const fontStyle = prop(options, 'font-style', style.getPropertyValue('font-style'));
  if (['normal', 'italic', 'oblique'].indexOf(fontStyle) !== -1) {
    font.push(fontStyle);
  }

  const fontVariant = prop(options, 'font-variant', style.getPropertyValue('font-variant'));
  if (['normal', 'small-caps'].indexOf(fontVariant) !== -1) {
    font.push(fontVariant);
  }

  const fontSize = prop(options, 'font-size', style.getPropertyValue('font-size')) || DEFAULTS['font-size'];
  const fontSizeValue = pxValue(fontSize);
  font.push(fontSizeValue + 'px');

  const fontFamily = prop(options, 'font-family', style.getPropertyValue('font-family')) || DEFAULTS['font-family'];
  font.push(fontFamily);

  return font.join(' ');
}

/**
 * Check for CSSStyleDeclaration
 *
 * @param val
 * @returns {bool}
 */
export function isCSSStyleDeclaration(val) {
  return val && typeof val.getPropertyValue === 'function';
}

/**
 * Check wether we can get computed style
 *
 * @param el
 * @returns {bool}
 */
export function canGetComputedStyle(el) {
  return isElement(el) && el.style && typeof window !== 'undefined' && typeof window.getComputedStyle === 'function';
}

/**
 * Check for DOM element
 *
 * @param el
 * @retutns {bool}
 */
export function isElement(el) {
  return typeof HTMLElement === 'object'
    ? el instanceof HTMLElement
    : Boolean(el && typeof el === 'object' && el !== null && el.nodeType === 1 && typeof el.nodeName === 'string');
}

/**
 * Check if argument is object
 * @param obj
 * @returns {boolean}
 */
export function isObject(obj) {
  return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
}

/**
 * Get style declaration if available
 *
 * @returns {CSSStyleDeclaration}
 */
export function getStyle(el, options) {
  const opts = {...(options || {})};
  const {style} = opts;
  if (!options) {
    options = {};
  }

  if (isCSSStyleDeclaration(style)) {
    return style;
  }

  if (canGetComputedStyle(el)) {
    return window.getComputedStyle(el, prop(options, 'pseudoElt', null));
  }

  return {
    getPropertyValue: key => prop(options, key),
  };
}

/**
 * Get styled text
 *
 * @param {string} text
 * @param {CSSStyleDeclaration} style
 * @returns {string}
 */
export function getStyledText(text, style) {
  switch (style.getPropertyValue('text-transform')) {
    case 'uppercase':
      return text.toUpperCase();
    case 'lowercase':
      return text.toLowerCase();
    default:
      return text;
  }
}

/**
 * Trim text and repace some breaking htmlentities for convenience
 * Point user to https://mths.be/he for real htmlentity decode
 * @param text
 * @returns {string}
 */
export function prepareText(text) {
  // Convert to unicode
  text = (text || '')
    .replace(/<wbr>/gi, '\u200B')
    .replace(/<br\s*\/?>/gi, '\u000A')
    .replace(/&shy;/gi, '\u00AD')
    .replace(/&mdash;/gi, '\u2014');

  if (/&#(\d+)(;?)|&#[xX]([a-fA-F0-9]+)(;?)|&([0-9a-zA-Z]+);/g.test(text) && console) {
    console.error(
      'text-metrics: Found encoded htmlenties. You may want to use https://mths.be/he to decode your text first.'
    );
  }

  return text.trim();
}

/**
 * Get textcontent from element
 * Try innerText first
 * @param el
 */
export function getText(el) {
  if (!el) {
    return '';
  }

  const text = el.innerText || el.textContent || '';

  return text.trim();
}

/**
 * Get property from src
 *
 * @param src
 * @param attr
 * @param defaultValue
 * @returns {*}
 */
export function prop(src, attr, defaultValue) {
  return (src && typeof src[attr] !== 'undefined' && src[attr]) || defaultValue;
}

/**
 * Normalize options
 *
 * @param options
 * @returns {*}
 */
export function normalizeOptions(options) {
  const opts = {};

  // Normalize keys (fontSize => font-size)
  Object.keys(options || {}).forEach(key => {
    const dashedKey = key.replace(/([A-Z])/g, $1 => '-' + $1.toLowerCase());
    opts[dashedKey] = options[key];
  });

  return opts;
}

/**
 * Get Canvas
 * @param font
 * @throws {Error}
 * @return {Context2d}
 */
export function getContext2d(font) {
  try {
    const ctx = document.createElement('canvas').getContext('2d');
    const dpr = window.devicePixelRatio || 1;
    const bsr =
      ctx.webkitBackingStorePixelRatio ||
      ctx.mozBackingStorePixelRatio ||
      ctx.msBackingStorePixelRatio ||
      ctx.oBackingStorePixelRatio ||
      ctx.backingStorePixelRatio ||
      1;
    ctx.font = font;
    ctx.setTransform(dpr / bsr, 0, 0, dpr / bsr, 0, 0);
    return ctx;
  } catch (error) {
    throw new Error('Canvas support required');
  }
}

/**
 * Check breaking character
 * http://www.unicode.org/reports/tr14/#Table1
 *
 * @param chr
 */
function checkBreak(chr) {
  return (
    (B2.includes(chr) && 'B2') ||
    (BAI.includes(chr) && 'BAI') ||
    (SHY.includes(chr) && 'SHY') ||
    (BA.includes(chr) && 'BA') ||
    (BB.includes(chr) && 'BB') ||
    (BK.includes(chr) && 'BK')
  );
}

export function computeLinesDefault({ctx, text, max, wordSpacing, letterSpacing}) {
  const addSpacing = addWordAndLetterSpacing(wordSpacing, letterSpacing);
  const lines = [];
  const parts = [];
  const breakpoints = [];
  let line = '';
  let part = '';

  if (!text) {
    return [];
  }

  // Compute array of breakpoints
  for (const chr of text) {
    const type = checkBreak(chr);
    if (part === '' && type === 'BAI') {
      continue;
    }

    if (type) {
      breakpoints.push({chr, type});

      parts.push(part);
      part = '';
    } else {
      part += chr;
    }
  }

  if (part) {
    parts.push(part);
  }

  // Loop over text parts and compute the lines
  for (let i = 0; i < parts.length; i++) {
    if (i === 0) {
      line = parts[i];
      continue;
    }

    const part = parts[i];
    if (BAI.includes(parts[i - 1]) && BAI.includes(parts[i])) {
      continue;
    }

    const breakpoint = breakpoints[i - 1];
    // Special treatment as we only render the soft hyphen if we need to split
    const chr = breakpoint.type === 'SHY' ? '' : breakpoint.chr;

    if (breakpoint.type === 'BK') {
      lines.push(line);
      line = part;
      continue;
    }

    // Measure width
    const rawWidth = ctx.measureText(line + chr + part).width + addSpacing(line + chr + part);
    const width = Math.round(rawWidth, 10);

    // Still fits in line
    if (width <= max) {
      line += chr + part;
      continue;
    }

    // Line is to long, we split at the breakpoint
    switch (breakpoint.type) {
      case 'SHY':
        lines.push(line + '-');
        line = part;
        break;
      case 'BA':
        lines.push(line + chr);
        line = part;
        break;
      case 'BAI':
        lines.push(line);
        line = part;
        break;
      case 'BB':
        lines.push(line);
        line = chr + part;
        break;
      case 'B2':
        if (parseInt(ctx.measureText(line + chr).width + addSpacing(line + chr), 10) <= max) {
          lines.push(line + chr);
          line = part;
        } else if (parseInt(ctx.measureText(chr + part).width + addSpacing(chr + part), 10) <= max) {
          lines.push(line);
          line = chr + part;
        } else {
          lines.push(line);
          lines.push(chr);
          line = part;
        }

        break;
      default:
        throw new Error('Undefoined break');
    }
  }

  if ([...line].length !== 0) {
    lines.push(line);
  }

  return lines;
}

export function computeLinesBreakAll({ctx, text, max, wordSpacing, letterSpacing}) {
  const addSpacing = addWordAndLetterSpacing(wordSpacing, letterSpacing);
  const lines = [];
  let line = '';
  let index = 0;

  if (!text) {
    return [];
  }

  for (const chr of text) {
    const type = checkBreak(chr);
    // Mandatory break found (br's converted to \u000A and innerText keeps br's as \u000A
    if (type === 'BK') {
      lines.push(line);
      line = '';
      continue;
    }

    const lineLength = line.length;
    if (BAI.includes(chr) && (lineLength === 0 || BAI.includes(line[lineLength - 1]))) {
      continue;
    }

    // Measure width
    let rawWidth = ctx.measureText(line + chr).width + addSpacing(line + chr);
    let width = Math.ceil(rawWidth);

    // Check if we can put char behind the shy
    if (type === 'SHY') {
      const next = text[index + 1] || '';
      rawWidth = ctx.measureText(line + chr + next).width + addSpacing(line + chr + next);
      width = Math.ceil(rawWidth);
    }

    // Needs at least one character
    if (width > max && [...line].length !== 0) {
      switch (type) {
        case 'SHY':
          lines.push(line + '-');
          line = '';
          break;
        case 'BA':
          lines.push(line + chr);
          line = '';
          break;
        case 'BAI':
          lines.push(line);
          line = '';
          break;
        default:
          lines.push(line);
          line = chr;
          break;
      }
    } else if (chr !== '\u00AD') {
      line += chr;
    }

    index++;
  }

  if ([...line].length !== 0) {
    lines.push(line);
  }

  return lines;
}
